engine/common/follower/cache/cache.go

```diff
@@ -0,0 +1,170 @@
+	}
+	distributor.AddConsumer(cache.handleEjectedEntity)
+	return cache
+}
+
+// handleEjectedEntity performs cleanup of secondary indexes to prevent memory leaks.
+// WARNING: Concurrency safety of this function is guaranteed by s.lock, this callback can be called
+// only in herocache.Cache.Add and we perform this call while s.lock is in locked state.
+func (c *Cache) handleEjectedEntity(entity flow.Entity) {
+	block := entity.(*flow.Block)
+	delete(c.byView, block.Header.View)
+	blocksByID := c.byParent[block.Header.ParentID]
+	delete(blocksByID, block.ID())
+	if len(blocksByID) == 0 {
+		delete(c.byParent, block.Header.ParentID)
+	}
+}
+
+// AddBlocks atomically applies batch of blocks to the cache of pending but not yet certified blocks. Upon insertion cache tries to resolve
+// incoming blocks to what is stored in the cache.
+// We require that incoming batch is sorted in ascending height order and doesn't have skipped blocks.
+// When receiving batch: [first, ..., last], we are only interested in first and last blocks. All blocks before
+// `last` are certified by construction (by the QC included in `last`).
+// Next scenarios are possible:
+// - for first block:
+//   - no parent available for first block.
+//   - parent for first block available in cache allowing to certify it, we can certify one extra block(parent).
+//
+// - for last block:
+//   - no child available for last block, need to wait for child to certify it.
+//   - child for last block available in cache allowing to certify it, we can certify one extra block(child).
+//
+// All blocks from the batch are stored in the cache to provide deduplication.
+// The function returns any new certified chain of blocks created by addition of the batch.
+// Returns `nil, nil` if the input batch has exactly one block and neither its parent nor child is in the cache.
+// Returns `certifiedBatch, certifyingQC` if the input batch has more than one block, and/or if either a child
+// or parent of the batch is in the cache.
+// Note that implementation behaves correctly where len(batch) == 1.
+// If message equivocation was detected it will be reported using a notification.
+// Concurrency safe.
+func (c *Cache) AddBlocks(batch []*flow.Block) (certifiedBatch []*flow.Block, certifyingQC *flow.QuorumCertificate) {
+	var equivocatedBlocks [][]*flow.Block
+
+	// prefill certifiedBatch with minimum viable result
+	// since batch is a chain of blocks, then by definition all except the last one
+	// has to be certified by definition
+	certifiedBatch = batch[:len(batch)-1]
+
+	if len(batch) > 1 {
+		// set certifyingQC, QC from last block certifies complete batch
+		certifyingQC = batch[len(batch)-1].Header.QuorumCertificate()
+	}
+	lastBlockID := batch[len(batch)-1].ID()
+
+	c.lock.Lock()
+	// check for message equivocation, report any if detected
+	for _, block := range batch {
+		if otherBlock, ok := c.byView[block.Header.View]; ok {
+			if otherBlock.ID() != block.ID() {
```

Comment:
ID computations are probably the most expensive operation by a large margin. I think it would be good to:
* precompute IDs for the input batch 
* change `byView` to `map[uint64]flow.Identifier` and store the block ID. In this case, you don't need to re-compute the ID in case there are repetitions.